crypto/dpop/dpop.go

```diff
@@ -0,0 +1,234 @@
+	}
+	// we require exactly one signature
+	if len(message.Signatures()) != 1 {
+		return nil, fmt.Errorf("%w: invalid number of signatures", ErrInvalidDPoP)
+	}
+	headers := message.Signatures()[0].ProtectedHeaders()
+	if !slices.Contains(SigningAlgValuesSupported, headers.Algorithm().String()) {
+		return nil, fmt.Errorf("%w: invalid alg: %s", ErrInvalidDPoP, headers.Algorithm())
+	}
+	if headers.Type() != "dpop+jwt" {
+		return nil, fmt.Errorf("%w: invalid type: %s", ErrInvalidDPoP, headers.Type())
+	}
+	if headers.JWK() == nil {
+		return nil, fmt.Errorf("%w: missing jwk header", ErrInvalidDPoP)
+	}
+	token, err := jwt.ParseString(s, jwt.WithKey(headers.Algorithm(), headers.JWK()))
+	if err != nil {
+		return nil, errors.Join(ErrInvalidDPoP, err)
+	}
+	if token.IssuedAt().Compare(time.Time{}) == 0 {
+		return nil, fmt.Errorf("%w: missing iat claim", ErrInvalidDPoP)
+	}
+	if v, ok := token.Get(HTUKey); !ok || v == "" {
+		return nil, fmt.Errorf("%w: missing htu claim", ErrInvalidDPoP)
+	}
+	if v, ok := token.Get(HTMKey); !ok || v == "" {
+		return nil, fmt.Errorf("%w: missing htm claim", ErrInvalidDPoP)
+	}
+	if token.JwtID() == "" {
+		return nil, fmt.Errorf("%w: missing jti claim", ErrInvalidDPoP)
+	}
+	if len(token.JwtID()) > maxJtiLength {
+		return nil, fmt.Errorf("%w: jti claim too long", ErrInvalidDPoP)
+	}
+
+	return &DPoP{raw: s, Token: token, Headers: headers}, nil
+}
+
+// HTU returns the htu claim of the DPoP token
+func (t DPoP) HTU() string {
+	if v, ok := t.Token.Get(HTUKey); ok {
+		return v.(string)
+	}
+	return ""
+}
+
+// HTM returns the htm claim of the DPoP token
+func (t DPoP) HTM() string {
+	if v, ok := t.Token.Get(HTMKey); ok {
+		return v.(string)
+	}
+	return ""
+}
+
+// Match checks if the JWK, http method, domain and path of the DPoP tokens match
+// for the url, the port is stripped.
+// If there is a mismatch, the reason is returned in an error.
+func (t DPoP) Match(jkt string, method string, url string) (bool, error) {
+	tp, _ := t.Headers.JWK().Thumbprint(crypto.SHA256)
+	base64tp := base64.RawURLEncoding.EncodeToString(tp)
+
+	if base64tp != jkt {
+		return false, errors.New("jkt mismatch")
+	}
+
+	// check method and url
+	if method != t.HTM() {
+		return false, errors.New("method mismatch")
+	}
+	urlLeft := strip(t.HTU())
+	urlRight := strip(url)
+	if urlLeft != urlRight {
+		return false, fmt.Errorf("url mismatch: %s, %s", urlLeft, urlRight)
+	}
+
+	return true, nil
+}
+
+func strip(raw string) string {
+	url, _ := url.Parse(raw)
+	url.Scheme = "https"
+	url.Host = strings.Split(url.Host, ":")[0]
+	return url.String()
+}
+
+func (t DPoP) MarshalJSON() ([]byte, error) {
+	quoted := fmt.Sprintf("%q", t.raw)
+	return []byte(quoted), nil
+}
+
+func (t *DPoP) UnmarshalJSON(bytes []byte) error {
+	if len(bytes) < 2 || bytes[0] != '"' || bytes[len(bytes)-1] != '"' {
+		return fmt.Errorf("invalid DPoP token: %s", string(bytes))
+	}
+	unquoted := string(bytes[1 : len(bytes)-1])
```

Comment:
could also do;

```go
var proofJWT string
if err := json.Unmarshal(bytes, &proofJWT); err != nil {
  return fmt.Errorf("invalid DPoP token: %s", string(bytes))
}
```

Reply:
marshalling overhead. AI suggested this (more basic) variant.